internals/secrethub/completion.go

```diff
@@ -0,0 +1,43 @@
+package secrethub
+
+import (
+	"os"
+
+	"github.com/secrethub/secrethub-cli/internals/cli"
+	// "github.com/spf13/cobra"
+)
+
+type CompletionCommand struct {
+	shell  cli.StringArgValue
+	clause *cli.CommandClause
+}
+
+// NewCompletionCommand is a command that, when executed, generates a completion script
+// for a specific shell, based on the argument it is provided with. It is able to generate
+// completions for Bash, ZSh, Fish and PowerShell.
+func NewCompletionCommand() *CompletionCommand {
+	return &CompletionCommand{}
+}
+
+// Register registers the command, arguments and flags on the provided Registerer.
+func (cmd *CompletionCommand) Register(r cli.Registerer) {
+	cmd.clause = r.Command("completion", "Generate completion script").Hidden()
+	cmd.clause.Cmd.DisableFlagsInUseLine = true
+	cmd.clause.Cmd.ValidArgs = []string{"bash", "zsh", "fish", "powershell"}
+	cmd.clause.BindAction(cmd.run)
+	cmd.clause.BindArguments([]cli.Argument{{Store: &cmd.shell, Name: "shell", Required: true}})
+}
+
+func (cmd *CompletionCommand) run() error {
+	switch cmd.shell.Param {
+	case "bash":
+		_ = cmd.clause.Cmd.Root().GenBashCompletion(os.Stdout)
```

Comment:
In order to be able to use completion, you need to install it. What the highlighted code does generates the code that, if put in a file in `/etc/bash_completions.d/`, it will enable you to use autocompletion for bash.
I will think of a way to automate this process. Although for bash you need sudo privileges.
What I have in mind is for the user to type `secrethub completion <shell>` and this will install autocompletion.
I don't think that cobra does this by default, compared to KingPin.

Reply:
Update: Now you can install autocompletion from CLI using the following command: `secrethub autocomplete bash`
It will ask you to confirm that you allow sudo privileges (answer `yes`, `y` or just press enter, just like apt install scenario), then you will need to insert your password for sudo and then follow the last step prompted.

Reply:
Sounds like an interesting approach. Two things to consider:
1. First of all, I think probably the most important change that still needs to be made here is not erroring if the completion is not configured. I expect that if I haven't configured completion then it just doesn't happen, not that I get an error.
2. Secondly, is this a common approach? Will our users find this command or will they look for something else: e.g. a script that configures completion. I suggest looking for what is the most common approach in other similar cli and choosing the one that makes the most sense.

Note that the current approach is very nice, possibly the simplest so let's not throw it away, but rather have a discussion once we find out what the most common approaches are.

Reply:
One interesting approach worth mentioning here is to have a shell script file for each type of shell which will be executed when autocompletion is asked (basically when making the double <TAB>). This is what I found at the `docker-cli`, which uses cobra as well. That approach will simplify things, since the user will no longer need to manually install autocompletion. However, more investigation is needed, since I can't figure out yet how the script is included in the binary file.

Reply:
Another observation: the current version of the cli has command completion without requiring any sort of configuration. Can we also achieve complete completion without requiring configuration if we use cobra? If not, can we at least keep the (sub)command completion working without any manual configuration and require manual config only for the path and other completions?

Reply:
Update:
I just looked through some cli's and here are my findings:
 * `aws-cli` - built in **python**, **does not support** autocompletion by default, you need to install it manually (link to the guide: https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-completion.html)
 * `docker-cli`- built in **go**, uses **cobra**, **supports** autocompletion by default for all types of shells. I think their approach is worth investigating, since they also provide completion for docker-container names and name of other services, which may come in handy for path-completion optimizations. From what I saw, they do not use the completion function provided by cobra, rather they made the completion script on their own. First on bash, then based on that one to the other shells. Not sure if we want to invest time in making our own cutom completion script, especially for the basic part, which can be easily generated with cobra. From here I will try to see how to include the completion script in the build process of the binary. 
 * `gcloud` - built in **python**, **does not support** autocompletion by default. It has however, an interaction shell that can be accessed by execution `gcloud alpha interactive`. There you have some nice autocompletion. (More details here: https://cloud.google.com/sdk/docs/interactive-gcloud)
 * `azure-cli` - built in **python**, **supports** autocompletion by default. The file "az.completion" is responsible for this. So I guess when the binary is built this part of code in in there as well, thus enabling autocompletion without manual installation
 * `kubectl` (kubernetes) - built in **go**, uses **cobra**, **does not support** autocompletion by default. You can generate the completion code by running `kubectl completion <shell>` for `bash` or `zsh`. By executing `kubectl completion -h` you get instructions of how to enbale autocompletion for kubernetes. These steps are the same ones that we do as well (e.g. enabling autocompletion for bash)
 * `terraform` - built in **go**, uses `github.com/mitchellh/cli` as their cli library, **does not support** autocompletion by default. Can be installed using `terraform -install-autocomplete` for `zsh` and `shell`. After that, you need to restart the shell to take effect.
 * `gh` (github cli) - built in **go**, uses **cobra**, **does not support** autocompletion by default. Completion scripts can be generated by executing `gh completion <shell>` for all types of shells.

Out of 7 CLI's that I looked into, only **2** support autocompletion by default. From the other 5, **4** will provide it by manuall installation, usually by sourcing the generated completion script. 

From this small investigation I can tell that it may not be a problem to provide autocompletion as an optional feature, thus letting the user to manually install it. However, having it by default would be a nice touch that may positively impact our users.

For now I will try to remove the error shown when no autocompletion is provided for our CLI. I propose having a discussion as to which version should we focus on: manuall installation or autocompletion enabled by default. It is worth saying that having it by default may take more time than manual installation, since only one cli from the list above is built in go, uses cobra as a cli and supports autocompletion by default. This means that I have to figure out how to include the script in the binary, which is quite challenging.
If we choose to go for manual installation, I think it would be nice for the user to install it by running a command instead of them doing all the work after the completion script has been generated. For that, I can look at `terraform` and see how they did it.

If you have any other examples of CLI's that I can investigate for some inspiration, please let me know. I hope this investigation can give you an overview of how autocompletion is percepted by both developers and users.

Reply:
Second Update:
After doing a bit more investigation for fixing the error with `unknown flag`, I realized that the completion script can be brought with the package installed from the package manager. This means we don't have to figure out how to include it in the binary, rather how to put it in the release package. I believe this solution may not be difficult, but I may need some guidance since release package creation is something new to me. 

A discussion based on these small discoveries should be held.